### PR TITLE
[Feature] Support YOLOv8 Ins Segmentation Inference

### PR DESCRIPTION
**Pull request by Nioolek**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation for this PR and the goal you want to achieve through this PR.

## Modification

Support YOLOv8 Ins Segmentation Inference

Can reproduce bbox and mask mAP on yolov8-s model.
image

Official mAP results:
image

### CODE REVIEW

File: configs/yolov8/yolov8_ins_s_syncbn_fast_8xb16-500e_coco.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
+_base_ = './yolov8_s_mask-refine_syncbn_fast_8xb16-500e_coco.py'
```
Comment (hhaAndroid) [inline configs/yolov8/yolov8_ins_s_syncbn_fast_8xb16-500e_coco.py:1]: 放到 config/yolov8/ins_seg 下

File: configs/yolov8/yolov8_ins_s_syncbn_fast_8xb16-500e_coco.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+# Batch size of a single GPU during validation
+val_batch_size_per_gpu = 16
+# Worker to pre-fetch data for each single GPU during validation
+val_num_workers = 8
+
+batch_shapes_cfg = dict(
```
Comment (hhaAndroid) [inline configs/yolov8/yolov8_ins_s_syncbn_fast_8xb16-500e_coco.py:8]: 我记得官方没有开这个

File: configs/yolov8/yolov8_ins_s_syncbn_fast_8xb16-500e_coco.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    _delete_=True,
+    type='BatchShapePolicy',
+    batch_size=val_batch_size_per_gpu,
+    img_size=_base_.img_scale[0],
+    # The image scale of padding should be divided by pad_size_divisor
+    size_divisor=32,
+    # Additional paddings for pixel scale
+    extra_pad_ratio=0.5)
+
+model_test_cfg = dict(
```
Comment (hhaAndroid) [inline configs/yolov8/yolov8_ins_s_syncbn_fast_8xb16-500e_coco.py:18]: 加个注释说明：推理非常慢，如果希望加速，可以增加 score_thr，并调小 nms_pre、max_per_img

File: mmyolo/datasets/transforms/transforms.py [hunk 4]
```diff
@@ -99,17 +99,18 @@ def _resize_img(self, results: dict):
                                             self.scale)
 
             if ratio != 1:
-                # resize image according to the ratio
-                image = mmcv.imrescale(
+                # resize image according to the shape
```
Comment (hhaAndroid) [inline mmyolo/datasets/transforms/transforms.py:102]: 实际上这个改法会导致目标检测掉点，这个掉点比较奇怪，理论上应该不会。你可以试一下。如果会掉点，目前能做的就是，增加一个 flag，检测任务和实例分割任务采用不同的做法?

File: mmyolo/models/dense_heads/yolov8_ins_head.py [hunk 5]
```diff
@@ -0,0 +1,468 @@
...
+        self.proto_preds = ProtoModule(
+            self.in_channels[0],
+            self.protos_channels,
+            self.masks_channels,
+            act_cfg=self.act_cfg,
+            norm_cfg=self.norm_cfg)
+
+        middle_channels = max(self.in_channels[0] // 4, self.masks_channels)
+        # mask coefficients preds
+        self.mask_coe_preds = nn.ModuleList(
```
Comment (hhaAndroid) [inline mmyolo/models/dense_heads/yolov8_ins_head.py:122]: mask_coe_preds -> mask_coeff_preds 或者干脆不要用简写

File: mmyolo/models/dense_heads/yolov8_ins_head.py [hunk 6]
```diff
@@ -0,0 +1,468 @@
...
+        if featmap_sizes != self.featmap_sizes:
+            self.mlvl_priors = self.prior_generator.grid_priors(
+                featmap_sizes,
+                dtype=cls_scores[0].dtype,
+                device=cls_scores[0].device,
+                with_stride=True)
+            self.featmap_sizes = featmap_sizes
+        flatten_priors = torch.cat(self.mlvl_priors)
+
+        mlvl_strides = [
```
Comment (hhaAndroid) [inline mmyolo/models/dense_heads/yolov8_ins_head.py:274]: 可以参考 rtmdet ins 那个pr，这个地方可以简化

File: mmyolo/models/dense_heads/yolov8_ins_head.py [hunk 7]
```diff
@@ -0,0 +1,468 @@
...
+                shape. Default to False.
+            mask_thr_binary (float): Threshold of mask prediction. Default
+                to 0.5.
+
+        Return:
+            Tensor: Instance segmentation masks for each instance.
+                Has shape (num_instance, H, W).
+        """
+
+        c, mh, mw = mask_proto.shape  # CHW
```
Comment (hhaAndroid) [inline mmyolo/models/dense_heads/yolov8_ins_head.py:428]: 这个地方也应该提供两个版本

**Comment by HRliao1109** [comment]
Hi @Nioolek 
Thanks in advance for your great work.
I'm trying to reproduce your work and make yolov8 instance segmentation trainable, but I encounter some issue when I'm working with it.
I clone the repo and download the weight from the ultralytics, but it come out to be an error message below

`Traceback (most recent call last):
  File "tools/test.py", line 150, in <module>
    main()
  File "tools/test.py", line 146, in main
    runner.test()
  File "/root/miniforge3/envs/mmyolo_v8_ins/lib/python3.8/site-packages/mmengine/runner/runner.py", line 1821, in test
    self.load_or_resume()
  File "/root/miniforge3/envs/mmyolo_v8_ins/lib/python3.8/site-packages/mmengine/runner/runner.py", line 1699, in load_or_resume
    self.load_checkpoint(self._load_from)
  File "/root/miniforge3/envs/mmyolo_v8_ins/lib/python3.8/site-packages/mmengine/runner/runner.py", line 2130, in load_checkpoint
    self.call_hook('after_load_checkpoint', checkpoint=checkpoint)
  File "/root/miniforge3/envs/mmyolo_v8_ins/lib/python3.8/site-packages/mmengine/runner/runner.py", line 1839, in call_hook
    getattr(hook, fn_name)(self, **kwargs)
  File "/root/miniforge3/envs/mmyolo_v8_ins/lib/python3.8/site-packages/mmengine/hooks/ema_hook.py", line 196, in after_load_checkpoint
    copy.deepcopy(checkpoint['state_dict']),
KeyError: 'state_dict'`
   
This error seems not cause by the difference between mmdetection and ultralytics instead of modified code. 
I wonder how you load the weight from the ultralytics or there is other way you can load it correctly?

Thanks in advance!

update

I found model yolov8_to_mmyolo.py in tools/model_converter. The weight downloaded in ultralytics should be convert to mmyolo by this tool.

**Comment by lijoe123** [comment]
Hello, Could you please tell me the process of yolov8 ins segmentation? I had try the yolov8 ins segmentation of your github, but it can't work. It occur the error : 
```
File "/home/music/Downloads/mmyolo-sam/mmyolo/models/dense_heads/yolov5_head.py", line 470, in loss
    losses = self.loss_by_feat(*loss_inputs)
TypeError: loss_by_feat() takes from 6 to 7 positional arguments but 8 were given
```
Could you please how to solve the problem? @Nioolek

**Comment by NickJony** [comment]
This code cannot be run because the function 'loss_by_feat' is not implemented.

**Comment by wangzixiang99** [comment]
> Hello, Could you please tell me the process of yolov8 ins segmentation? I had try the yolov8 ins segmentation of your github, but it can't work. It occur the error :
> 
> ```
> File "/home/music/Downloads/mmyolo-sam/mmyolo/models/dense_heads/yolov5_head.py", line 470, in loss
>     losses = self.loss_by_feat(*loss_inputs)
> TypeError: loss_by_feat() takes from 6 to 7 positional arguments but 8 were given
> ```
> 
> Could you please how to solve the problem? @Nioolek

hello,请问这个问题解决了嘛？我也遇到了同样的问题

**Comment by FreeWillThorn** [comment]
> > Hello, Could you please tell me the process of yolov8 ins segmentation? I had try the yolov8 ins segmentation of your github, but it can't work. It occur the error :
> > ```
> > File "/home/music/Downloads/mmyolo-sam/mmyolo/models/dense_heads/yolov5_head.py", line 470, in loss
> >     losses = self.loss_by_feat(*loss_inputs)
> > TypeError: loss_by_feat() takes from 6 to 7 positional arguments but 8 were given
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Could you please how to solve the problem? @Nioolek
> 
> hello,请问这个问题解决了嘛？我也遇到了同样的问题

hi , have you figured it out?
（解决了吗？我看他后来也妹更新啊）

**Pull request merged**